censusdis/data.py

```diff
@@ -1246,3 +1258,180 @@ def geographies(dataset: str, vintage: VintageType) -> List[List[str]]:
+
+    Parameters
+    ----------
+    gdf_geo
+        A GeoDataFrame containing polygons within the United States
+    gdf_water
+        A GeoDataFrame containing census AREAWATER polygons
+    minimum_area_sq_meters
+        The smallest water polygon to be removed, specified in square meters
+
+    Returns
+    -------
+        A version of gdf_geo with the water areas removed
+    """
+
+    # Combining polygons speeds up the overlay operation
+    geo_combined_water = gdf_water[
+        gdf_water["AWATER"] >= minimum_area_sq_meters
+    ].unary_union
+    gdf_combined_water = gpd.GeoDataFrame(geometry=[geo_combined_water])
+    gdf_combined_water = gdf_combined_water.set_crs(gdf_water.crs)
+
+    return gdf_geo.overlay(
+        gdf_combined_water,
+        "difference",
+        keep_geom_type=False,
+    )
+
+
+def remove_slivers(gdf_geo: gpd.GeoDataFrame, threshhold: Number):
+    """
+    Removes slivers by identifying components of geometries that have a low area
+    relative to the full geometry. First, separates multipolygon and geometry
+    collections into component pieces. Then calculates area proportion of each
+    component. Components falling below a specified threshold are dropped.
+
+    Parameters
+    ----------
+    gdf_geo
+        A GeoDataFrame containing polygons within the United States
+    threshold
+        The minimum proportion of the full geometry area that each component must
+        have to be preserved.
+
+    Returns
+    -------
+        A version of gdf_geo that has the small components removed
+    """
+
+    # We need to convert between the input CRS and a planar crs to calculate area
+    input_crs = gdf_geo.crs
+    gdf_geo = gdf_geo.to_crs(5070)  # Albers Equal Area
+
+    # Any slivers will be found in geometries of type 'Multipolygon' or 'Geometry Collection'
+    # After exploding the geomtries we'll need to re-assemble them using these columns
+    group_columns = gdf_geo.columns.to_list()
+    group_columns.remove("geometry")
+
+    # We split these non-polygon geometries up and identify the area of each component
+    gdf_geo.loc[:, "full_area"] = gdf_geo.area
+    gdf_exploded = gdf_geo.explode(index_parts=True)
+    gdf_exploded.loc[:, "component_area"] = gdf_exploded.area
+    gdf_exploded.loc[:, "area_pct"] = (
+        gdf_exploded.component_area / gdf_exploded.full_area
+    )
+
+    # Component geometry that make up a percentage of the full geometry area below the
+    # specified threshold are removed
+    geometries_to_include = gdf_exploded[gdf_exploded.area_pct >= threshhold]
+
+    # Back to original crs
+    geometries_to_include = geometries_to_include.to_crs(input_crs)
+
+    # Geopandas.dissolve doesn't like missing values. We need to replace them to perform the operation
+    geometries_to_include[group_columns] = geometries_to_include[group_columns].fillna(
+        -9999
+    )
+    gdf_out = geometries_to_include.dissolve(group_columns).reset_index()
```

Comment:
I'm concerned about the exploding and dissolving and what that will do to the values in group columns. They could be totals, or medians, or whatever, and we don't really know how to put them back together. I'd rather try to work strictly on the geometries. See my subsequent commit for a way of doing this with separate geo-only operations with their own small independent unit tests.